### PR TITLE
Add support for cross-host live migration over TCP

### PR DESCRIPTION
**Pull request by ljrcore**
# Summary

This pull request updates the vm-memory crate from version 0.16.0 to 0.16.1, incorporating the necessary implementations for TcpStream. It introduces TCP-based cross-host live migration support in Cloud Hypervisor, enhancing migration performance and simplifying the process. Additionally, the live migration documentation has been updated to provide instructions for executing cross-host live migrations over TCP connections.

# Changes Made

## `build`: Bump `vm-memory` from `0.16.0` to `0.16.1`

- **Updated** the `vm-memory` crate dependency to version `0.16.1` to include the implementations of `ReadVolatile` and `WriteVolatile` for `TcpStream`, which are necessary for TCP-based live migration.

## `vmm`: Add support for cross-host live migration over TCP

- **Modified** `vm_send_migration` and `vm_receive_migration` functions to support `destination_url` and `receiver_url` starting with `tcp:`, allowing live migration over TCP sockets.
- **Moved** socket connection logic into `vm_send_migration` and `vm_receive_migration`, simplifying the migration functions.
- **Improved** migration efficiency by implementing direct TCP connections between hosts to transmit data.

## `docs`: Add documentation for cross-host TCP live migration

- **Updated** the live migration documentation to include instructions for performing cross-host live migrations over TCP connections.
- **Provided** detailed steps, commands, and considerations for users to perform TCP-based cross-host migrations.

# Test

We conducted tests to compare the performance of the new TCP-based cross-host live migration with the existing Unix socket migration using `socat` for cross-host communication. The key metrics measured were the **Total Migration Time** and **Downtime** experienced during the migration process.

## **Testing Environment**

- **Host Machines**: Both the source and destination hosts are equipped with identical hardware specifications.

- Virtual Machine Under Test:

  - **CPU**: 4 vCPUs
  - **Memory**: 8 GB RAM

- Memory Pressure Application

  - Inside the guest VM, we applied memory pressure using the following command:

    ```bash
    stress -m 4 --vm-bytes 256M
    ```

### Estimating Migration Time and Downtime from Logs

The migration time and downtime were estimated using timestamps from the Cloud Hypervisor logs. The methodology is as follows:

- **Migration Time**: Calculated as the difference between the timestamps when the migration started and when it completed.
  - **Migration Start Time**: The timestamp when the `VmSendMigration` API request is received.
  - **Migration End Time**: The timestamp when the migration is marked as complete (`Migration complete` log entry).
- **Downtime**: Calculated as the difference between the timestamps when the VM is paused for migration and when the migration is complete.
  - **Downtime Start Time**: The timestamp when the VM is paused (`VM has been paused` log entry).
  - **Downtime End Time**: The same as the Migration End Time.

### TCP Migration Example

```
Start Migration: cloud-hypervisor: 12.852263s:  INFO:vmm/src/api/mod.rs:1262 -- API request event: VmSendMigration VmSendMigrationData { destination_url: "tcp:192.168.102.2:6000", local: false }
Start Downtime: cloud-hypervisor: 17.796833s: INFO:vmm/src/vm.rs:2488 -- VM has been paused
End Migration: cloud-hypervisor: 18.226345s:  INFO:vmm/src/lib.rs:1167 -- Migration complete
```

**Calculations:**

- **Migration Time** = Migration End Time - Migration Start Time
  - **Migration Time** = 18.226345s - 12.852263s = **5374.082 ms**
- **Downtime** = Migration End Time - Downtime Start Time
  - **Downtime** = 18.226345s - 17.796833s = **429.512 ms**

### TCP Migration Results

| Test        | **Migration Time (ms)** | **Downtime (ms)** |
| ----------- | ----------------------- | ----------------- |
| 1           | 5374.082                | 429.512           |
| 2           | 5347.307                | 353.986           |
| 3           | 5364.191                | 404.331           |
| **Average** | **5361.860**            | **395.943**       |

### Unix Socket Migration with Socat Results

| Test        | **Migration Time (ms)** | **Downtime (ms)** |
| ----------- | ----------------------- | ----------------- |
| 1           | 11797.579               | 845.931           |
| 2           | 11211.548               | 786.769           |
| 3           | 11689.248               | 790.308           |
| **Average** | **11566.125**           | **807.669**       |

## Analysis

- **Migration Time:**
  - TCP migration shows significantly faster migration times, averaging approximately **5,362 ms**.
  - Unix socket migration with `socat` takes longer, averaging approximately **11,566 ms**.
  - TCP migration reduces the total migration time by about **53%** compared to Unix socket migration.
- **Downtime:**
  - Downtime during TCP migration is shorter, averaging approximately **396 ms**.
  - Downtime during Unix socket migration averages around **808 ms**.
  - TCP migration reduces downtime by about **51%** compared to Unix socket migration.

The implementation of TCP-based cross-host live migration significantly improves migration performance when compared to Unix socket migration using `socat`. By reducing both the total migration time and the downtime by over 50%, this enhancement increases the efficiency and reliability of live migrations in cross-host scenarios.

### CODE REVIEW

File: vmm/src/vm.rs [hunk 1]
```diff
@@ -2440,6 +2485,7 @@ impl Vm {
 impl Pausable for Vm {
     fn pause(&mut self) -> std::result::Result<(), MigratableError> {
         event!("vm", "pausing");
+        info!("Pause VM for migration");
```
Comment (up2wing) [inline vmm/src/vm.rs:2488]: This log may not be right, 'cause vm may pause from api request also.

Comment (ljrcore) [reply vmm/src/vm.rs]: The new version has fixed this problem.

Thanks

**Review by rbradford — COMMENTED** [review]
Please add integration tests.

**Comment by ljrcore** [comment]
> Please add integration tests.

Integration tests have been added in the new pushed PR:)

Thanks

**Comment by TimePrinciple** [comment]
@ljrcore Would you mind rebasing to capture the new clippy fix?

**Comment by ljrcore** [comment]
> @ljrcore Would you mind rebasing to apply the new clippy fix?

Done.

Thanks

**Comment by ljrcore** [comment]
@blitz The live migration documentation has been updated.

File: vmm/src/lib.rs [hunk 2]
```diff
@@ -997,7 +1079,14 @@ impl Vmm {
         };
 
         if send_data_migration.local {
-            vm.send_memory_fds(&mut socket)?;
+            match &mut socket {
+                SocketStream::Unix(unix_socket) => {
+                    vm.send_memory_fds(unix_socket)?;
+                }
+                SocketStream::Tcp(tcp_socket) => {
+                    vm.send_memory_fds_tcp(tcp_socket)?;
```
Comment (rbradford) [inline vmm/src/lib.rs:1087]: The purpose of `--local` is to not need to read/write the memory and instead use SCM_RIGHTS to move the fd across the process boundaries. `--local` makes no sense with TCP and should generate an error.

File: vmm/src/lib.rs [hunk 3]
```diff
@@ -1081,6 +1170,196 @@ impl Vmm {
...
+                            }
+
+                            Response::ok().write_to(&mut socket).map_err(|e| {
+                                MigratableError::MigrateReceive(anyhow!(
+                                    "Error sending response: {}",
+                                    e
+                                ))
+                            })?;
+                        }
+                        SocketStream::Tcp(tcp_socket) => {
```
Comment (rbradford) [inline vmm/src/lib.rs:1275]: This case should be an error - `MemoryFd` command should not be sent over TCP.

File: vmm/src/lib.rs (hunk 2 above)
Comment (ljrcore) [reply vmm/src/lib.rs]: You are right. I have modified the code as you suggested to ensure that when using TCP, if the user attempts to use the --local option, they will get an error message:)

Thanks

File: vmm/src/lib.rs (hunk 3 above)
Comment (ljrcore) [reply vmm/src/lib.rs]: > This case should be an error - `MemoryFd` command should not be sent over TCP.

Thanks for pointing out this issue. I have modified the code to ensure that an error message is generated in this case.

File: vmm/src/vm.rs [hunk 4]
```diff
@@ -2439,6 +2439,7 @@ impl Vm {
 
 impl Pausable for Vm {
     fn pause(&mut self) -> std::result::Result<(), MigratableError> {
+        info!("VM has been paused");
```
Comment (TimePrinciple) [inline vmm/src/vm.rs:2442]: IMHO here is not the right place for this and contradicts just the line below, would line 2476 and lines after be more appropriate 🧐

File: vmm/src/lib.rs [hunk 5]
```diff
@@ -1081,6 +1173,149 @@ impl Vmm {
...
+                            );
+                            Response::error().write_to(&mut socket).map_err(|e| {
+                                MigratableError::MigrateReceive(anyhow!(
+                                    "Error sending response: {}",
+                                    e
+                                ))
+                            })?;
+                            return Err(MigratableError::MigrateReceive(anyhow!(
+                                "MemoryFd command not supported over TCP sockets"
+                            )));
```
Comment (TimePrinciple) [inline vmm/src/lib.rs:1291]: Since `send_memory_fd` would (and should) never be invoked on `TcpSocket`, should we change the whole section to `unreachable!("MemoryFd command not supported over TCP sockets")`

Comment (ljrcore) [reply vmm/src/lib.rs:1291]: > Since `send_memory_fd` would (and should) never be invoked on `TcpSocket`, should we change the whole section to `unreachable!("MemoryFd command not supported over TCP sockets")`

You are right. TCP does not use MemoryFd, so I will remove the TCP related code here.

File: vmm/src/lib.rs (hunk 3 above)
Comment (ljrcore) [reply vmm/src/lib.rs]: > This case should be an error - `MemoryFd` command should not be sent over TCP.

Yes, TCP does not use MemoryFd. The code has been updated:)

Thanks

File: vmm/src/lib.rs [hunk 6]
```diff
@@ -2101,11 +2236,29 @@ impl RequestHandler for Vmm {
...
+            SocketStream::Tcp(socket)
+        } else {
+            let path = Self::socket_url_to_path(&send_data_migration.destination_url)?;
+            info!("Connecting to UNIX socket at {:?}", path);
+            let socket = UnixStream::connect(path).map_err(|e| {
+                MigratableError::MigrateSend(anyhow!("Error connecting to UNIX socket: {}", e))
+            })?;
+            SocketStream::Unix(socket)
+        };
+
```
Comment (rbradford) [inline vmm/src/lib.rs:2255]: Please extract this into a method that returns `Result<SocketStream,...>`

File: vmm/src/lib.rs [hunk 7]
```diff
@@ -1958,124 +2176,41 @@ impl RequestHandler for Vmm {
...
+        // Accept the connection and get the socket
+        let socket = match listener {
+            SocketListener::Unix(listener, path) => {
+                let (socket, _addr) = listener.accept().map_err(|e| {
+                    MigratableError::MigrateReceive(anyhow!("Error unlinking UNIX socket: {}", e))
+                })?;
+                std::fs::remove_file(path).map_err(|e| {
+                    MigratableError::MigrateReceive(anyhow!("Error accepting on TCP socket: {}", e))
+                })?;
+                SocketStream::Unix(socket)
```
Comment (rbradford) [inline vmm/src/lib.rs:2203]: Ditto, I think this could be extracted to a function.

File: vmm/src/lib.rs [hunk 8]
```diff
@@ -1958,124 +2176,41 @@ impl RequestHandler for Vmm {
...
-                    self.vm = None;
-                    self.vm_config = None;
-                    Response::ok().write_to(&mut socket).ok();
-                    break;
-                }
+        // Accept the connection and get the socket
+        let socket = match listener {
+            SocketListener::Unix(listener, path) => {
+                let (socket, _addr) = listener.accept().map_err(|e| {
+                    MigratableError::MigrateReceive(anyhow!("Error unlinking UNIX socket: {}", e))
```
Comment (rbradford) [inline vmm/src/lib.rs:2198]: Not sure about this error message?

**Review by rbradford — COMMENTED** [review]
Looking good - just some refactoring I think would be helpful!

**Comment by TimePrinciple** [comment]
> Hi, @rbradford
> 
> A check failed, how can I fix it?

It could be flakes, you could pin me to restart the failed job to find out

**Comment by ljrcore** [comment]
> > Hi, @rbradford
> > A check failed, how can I fix it?
> 
> It could be flakes, you could pin me to restart the failed job to find out

Got it, thanks:)

**Comment by ljrcore** [comment]
> Looking good - just some refactoring I think would be helpful!

Thanks for your suggestion, the code has been updated.

File: vmm/src/lib.rs [hunk 9]
```diff
@@ -1958,124 +2232,10 @@ impl RequestHandler for Vmm {
...
-                    self.vm = None;
-                    self.vm_config = None;
-                    Response::ok().write_to(&mut socket).ok();
-                    break;
-                }
-            }
-        }
-
-        Ok(())
+        self.receive_migration(socket)
```
Comment (likebreath) [inline vmm/src/lib.rs:2238]: Let's not extract this into a new function, otherwise the `vm_receive_migration` function will be a trivial wrapper that contains merely two function calls. Also, keeping the original function will reduce the change size, making the real change clear (such as the changes needed for `Command::MemoryFd`).

File: tests/integration.rs [hunk 10]
```diff
@@ -10295,6 +10297,207 @@ mod live_migration {
...
+            .arg(pmem_temp_file.as_path())
+            .output()
+            .expect("Expect creating disk image to succeed");
+        let pmem_path = String::from("/dev/pmem0");
+        // Start the source VM
+        let src_vm_path = if !upgrade_test {
+            clh_command("cloud-hypervisor")
+        } else {
+            cloud_hypervisor_release_path()
+        };
```
Comment (likebreath) [inline tests/integration.rs:10403]: Live upgrade won't work since the older version of CH simply does not support TCP socket. Also, I believe we won't need this test, as live upgrade has been well tested with other tests (using Unix socket of course). I'd just drop it from the new test.

**Review by likebreath — COMMENTED** [review]
Thank you for the contribution. Generally looking good to me. Some comments below.

One note, I can see that using TCP socket can reduce the live migration downtime over using Unix socket, but I won't draw the conclusion that it reduces the downtime by 50%. The live migration downtime depends a lot on the workload being running.

**Comment by ljrcore** [comment]
> Thank you for the contribution. Generally looking good to me. Some comments below.
> 
> One note, I can see that using TCP socket can reduce the live migration downtime over using Unix socket, but I won't draw the conclusion that it reduces the downtime by 50%. The live migration downtime depends a lot on the workload being running.

Thanks for your review. The live migration downtime does depend a lot on the workload being run. I got this data when I applied memory pressure using the following command. 

```
stress -m 4 --vm-bytes 256M
```

If the memory pressure changes, the data will change accordingly. Maybe it is also related to the test environment. I used a 4c8g virtual machine for testing.

File: vmm/src/lib.rs [hunk 11]
```diff
@@ -1081,6 +1229,132 @@ impl Vmm {
...
+                        Response::error().write_to(&mut socket).map_err(|e| {
+                            MigratableError::MigrateReceive(anyhow!(
+                                "Error sending response: {}",
+                                e
+                            ))
+                        })?;
+                        continue;
+                    }
+
+                    if let SocketStream::Unix(unix_socket) = &mut socket {
```
[the remark this answers is not in the document]
Comment (ljrcore) [reply vmm/src/lib.rs:1308]: > The `MemoryFd` command is used with local migration only, and hence won't work with the tcp socket. Let's make sure we report an error and abort the migration for that case.

Yes, the MemoryFd command is only used for local migration. There is no need for any tcp processing here, because tcp live migration will never call the MemoryFd command.

File: vmm/src/lib.rs (hunk 9 above)
Comment (ljrcore) [reply vmm/src/lib.rs:2238]: > Let's not extract this into a new function, otherwise the `vm_receive_migration` function will be a trivial wrapper that contains merely two function calls. Also, keeping the original function will reduce the change size, making the real change clear (such as the changes needed for `Command::MemoryFd`).

I will make the changes as you suggested, thank you for your feedback.

File: tests/integration.rs (hunk 10 above)
Comment (ljrcore) [reply tests/integration.rs:10403]: > Live upgrade won't work since the older version of CH simply does not support TCP socket. Also, I believe we won't need this test, as live upgrade has been well tested with other tests (using Unix socket of course). I'd just drop it from the new test.

I originally wanted to test tcp live migration over the local network. Maybe this is not needed, I will remove it, thanks.

Comment (rbradford) [reply tests/integration.rs]: You should keep the integration test just test over "127.0.0.1" the `--local migration has a very specific meaning - the memory is not copied and instead shared by fd.

File: vmm/src/vm.rs [hunk 12]
```diff
@@ -2472,6 +2472,7 @@ impl Pausable for Vm {
         self.vm
             .pause()
             .map_err(|e| MigratableError::Pause(anyhow!("Could not pause the VM: {}", e)))?;
+        info!("VM has been paused");
```
Comment (likebreath) [inline vmm/src/vm.rs:2475]: Let's drop this change, since this is not related to the topic.

File: vmm/src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (likebreath) [reply vmm/src/lib.rs]: Maybe I was not clear about the changes that I suggested earlier. It is better to use `match` instead of `if let` here, so that you can explicitly report errors when TCP socket is used with local migration (like what you have done from `send_migration`).

File: docs/live_migration.md [hunk 13]
```diff
@@ -75,7 +72,11 @@ src $ curl $DEBIAN/initrd.gz > /var/images/initrd
 
 Repeat the above steps on the destination host.
 
-### Starting the Receiver VM
+### Unix Socket Migration
+
+If TCP socket is selected for migration, we can tunnel traffic through "socat".
```
Comment (likebreath) [inline docs/live_migration.md:77]: Typo: `TCP socket` -> `Unix socket`

File: docs/live_migration.md [hunk 14]
```diff
@@ -127,3 +128,54 @@ src $ ch-remote --api-socket=/tmp/api send-migration unix:/tmp/sock
 
 When the above commands completed, the VM should be successfully
 migrated to the destination machine without interrupting the workload.
+
+### TCP Socket Migration
+
+If TCP socket is selected for migration, the cross-host migration
+performance will be further improved, but it is necessary to consider
```
Comment (likebreath) [inline docs/live_migration.md:135]: I'd drop the comment on performance differences, since we don't have much data to back it up. Also, the performance differences depend a lot on the end user environment. Let's leave it to users to evaluate and choose what's the best for their use cases.

**Review by likebreath — COMMENTED** [review]
I believe you accidentally dropped the commit that adds an integration test. 

I took a closer look at the PR and have some additional comments below.

File: vmm/src/lib.rs [hunk 15]
```diff
@@ -2031,28 +2171,40 @@ impl RequestHandler for Vmm {
 
                     if !started {
                         warn!("Migration not started yet");
-                        Response::error().write_to(&mut socket)?;
+                        Response::error().write_to(&mut socket).map_err(|e| {
+                            MigratableError::MigrateReceive(anyhow!(
+                                "Error sending response: {}",
+                                e
+                            ))
+                        })?;
```
[the remark this answers is not in the document]
Comment (ljrcore) [reply vmm/src/lib.rs:2179]: > I don't see why this change is needed.

This is to ensure that any errors that occur while writing to the socket are properly caught and handled, instead of causing a panic.

File: vmm/src/lib.rs [hunk 16]
```diff
@@ -2031,28 +2171,40 @@ impl RequestHandler for Vmm {
...
+                        let mut buf = [0u8; 4];
+                        let (_, file) = unix_socket.recv_with_fd(&mut buf).map_err(|e| {
+                            MigratableError::MigrateReceive(anyhow!(
+                                "Error receiving slot from socket: {}",
+                                e
+                            ))
+                        })?;
+
+                        if existing_memory_files.is_none() {
+                            existing_memory_files = Some(HashMap::new());
```
[the remark this answers is not in the document]
Comment (ljrcore) [reply vmm/src/lib.rs:2193]: > Ditto. Don't see why this change is needed.
> 
> Although HashMap::new() works here too, HashMap::default() is more flexible in general, e.g. support generic types that are determined at compile time.

I agree with your suggestion and it will not be modified here.

File: vmm/src/lib.rs [hunk 17]
```diff
@@ -2031,28 +2171,40 @@ impl RequestHandler for Vmm {
...
-                    if let Some(ref mut existing_memory_files) = existing_memory_files {
-                        let slot = u32::from_le_bytes(buf);
-                        existing_memory_files.insert(slot, file.unwrap());
+                        Response::ok().write_to(&mut socket).map_err(|e| {
+                            MigratableError::MigrateReceive(anyhow!(
+                                "Error sending response: {}",
+                                e
+                            ))
+                        })?;
                     }
```
[the remark this answers is not in the document]
Comment (ljrcore) [reply vmm/src/lib.rs:2207]: > Ditto. Don't see why this change is needed.

Again, this is to ensure that any errors that occur when writing to the socket are properly caught and handled. I am not sure if we need such error handling, if not, the next commit will cancel this change.

File: vmm/src/lib.rs [hunk 18]
```diff
@@ -2101,11 +2253,15 @@ impl RequestHandler for Vmm {
...
 
+        // Set up the socket connection
+        let socket = Vmm::connect_to_migration_socket(&send_data_migration.destination_url)?;
+
         if let Some(vm) = self.vm.as_mut() {
             Self::send_migration(
                 vm,
                 #[cfg(all(feature = "kvm", target_arch = "x86_64"))]
                 self.hypervisor.clone(),
+                socket,
```
[the remark this answers is not in the document]
Comment (ljrcore) [reply vmm/src/lib.rs:2264]: > send_migraiton()

Ok, I will move it to `send_migraiton()`.

[the remark this answers is not in the document]
Comment (ljrcore) [reply vmm/src/lib.rs:2264]: > It would make more sense to call this function within `send_migraiton()` which also introduces less code changes.

Ok, I will move it to `send_migraiton()`.

File: vmm/src/vm.rs (hunk 12 above)
Comment (ljrcore) [reply vmm/src/vm.rs:2475]: > Let's drop this change, since this is not related to the topic.

This is to estimate downtime through logging. It may not be needed here, I will remove it.

File: docs/live_migration.md (hunk 14 above)
Comment (ljrcore) [reply docs/live_migration.md:135]: > I'd drop the comment on performance differences, since we don't have much data to back it up. Also, the performance differences depend a lot on the end user environment. Let's leave it to users to evaluate and choose what's the best for their use cases.

You are right, I will remove the description of performance here.

**Comment by ljrcore** [comment]
> Maybe I was not clear about the changes that I suggested earlier. It is better to use match instead of if let here, so that you can explicitly report errors when TCP socket is used with local migration (like what you have done from send_migration).

I don't quite understand this description. I can change the `if let` to `match`, but there will be no handling of TCP here. If TCP socket is used with local migration, it will report an error in `if send_data_migration.local`.

File: vmm/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (rbradford) [reply vmm/src/lib.rs:2179]: The `?` should propagate the error - was something at a higher level not handling this correctly?

File: vmm/src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ljrcore) [reply vmm/src/lib.rs]: > Maybe I was not clear about the changes that I suggested earlier. It is better to use `match` instead of `if let` here, so that you can explicitly report errors when TCP socket is used with local migration (like what you have done from `send_migration`).

I don't quite understand this description. I can change the `if let` to `match`, but there will not be any handling of TCP here, because the handling of TCP here will be redundant. If TCP sockets are used with local migration, an error will be reported in `if send_data_migration.local`.

```
        if send_data_migration.local {
            match &mut socket {
                SocketStream::Unix(unix_socket) => {
                    // Proceed with sending memory file descriptors over UNIX socket
                    vm.send_memory_fds(unix_socket)?;
                }
                SocketStream::Tcp(_tcp_socket) => {
                    return Err(MigratableError::MigrateSend(anyhow!(
                        "--local option is not supported with TCP sockets",
                    )));
                }
            }
        }
```

**Comment by ljrcore** [comment]
Thanks for your patience and careful review of my code. I have modified and updated the code according to your suggestions. I appreciate your time and effort in this process.

**Comment by ljrcore** [comment]
> > Hi, @rbradford
> > A check failed, how can I fix it?
> 
> It could be flakes, you could pin me to restart the failed job to find out

@TimePrinciple Hi, Ruoqing

Could you restart failed jobs?

File: vmm/src/lib.rs [hunk 19]
```diff
@@ -919,14 +1003,72 @@ impl Vmm {
...
+
+            let socket = UnixStream::connect(&path).map_err(|e| {
+                MigratableError::MigrateSend(anyhow!("Error connecting to UNIX socket: {}", e))
+            })?;
+
+            Ok(SocketStream::Unix(socket))
+        }
+    }
+
+    fn receive_migraiton_socket(
```
Comment (rbradford) [inline vmm/src/lib.rs:1029]: Typo of the method name

**Review by likebreath — APPROVED** [review]

File: vmm/src/lib.rs [hunk 20]
```diff
@@ -2035,24 +2177,38 @@ impl RequestHandler for Vmm {
...
+                        }
+                        SocketStream::Tcp(_tcp_socket) => {
+                            // For TCP sockets, we cannot transfer file descriptors
+                            warn!(
+                                "MemoryFd command received over TCP socket, which is not supported"
+                            );
+                            Response::error().write_to(&mut socket)?;
+                            return Err(MigratableError::MigrateReceive(anyhow!(
+                                "MemoryFd command not supported over TCP sockets"
+                            )));
```
Comment (likebreath) [inline vmm/src/lib.rs:2209]: Let's drop the `return Err()`.

I don't think we will need to abort the destination VM explicitly here, since it is mostly driven by the source VM. Let's leave it up to the source VM to decide what to do when there is an error returned from the destination VM side. As for now, it will send the "Abandon" request to stop the destination VM.

**Review by likebreath — CHANGES_REQUESTED** [review]
Almost forgot this last one issue. Other than this, I think this PR is ready to be landed.

File: vmm/src/lib.rs (hunk 20 above)
Comment (ljrcore) [reply vmm/src/lib.rs]: > Let's drop the `return Err()`.
> 
> I don't think we will need to abort the destination VM explicitly here, since it is mostly driven by the source VM. Let's leave it up to the source VM to decide what to do when there is an error returned from the destination VM side. As for now, it will send the "Abandon" request to stop the destination VM.

I have drop the `return Err()`, thanks.

**Review by likebreath — APPROVED** [review]

**Review by rbradford — APPROVED** [review]
